homeassistant/components/bayesian/binary_sensor.py

```diff
@@ -380,82 +393,86 @@ def _build_observations_by_template(self):
         """Return True if numeric condition is met, return False if not, return None otherwise."""
-        entity = entity_observation["entity_id"]
+        entity = entity_observation.entity_id
 
         try:
             if condition.state(self.hass, entity, [STATE_UNKNOWN, STATE_UNAVAILABLE]):
                 return None
             return condition.async_numeric_state(
                 self.hass,
                 entity,
-                entity_observation.get("below"),
-                entity_observation.get("above"),
+                entity_observation.below,
+                entity_observation.above,
                 None,
-                entity_observation,
+                entity_observation.to_dict(),
             )
         except ConditionError:
             return None
 
     def _process_state(self, entity_observation):
-        """Return True if state conditions are met."""
-        entity = entity_observation["entity_id"]
+        """Return True if state conditions are met, return False if they are not.
+
+        Returns None if the state is unavailable.
+        """
+
+        entity = entity_observation.entity_id
 
         try:
             if condition.state(self.hass, entity, [STATE_UNKNOWN, STATE_UNAVAILABLE]):
                 return None
 
-            return condition.state(
-                self.hass, entity, entity_observation.get("to_state")
-            )
+            return condition.state(self.hass, entity, entity_observation.to_state)
         except ConditionError:
             return None
 
     def _process_multi_state(self, entity_observation):
-        """Return True if state conditions are met."""
-        entity = entity_observation["entity_id"]
+        """Return True if state conditions are met, otherwise return None.
+
+        Never return False as all other states should have their own probabilities configured.
+        """
+
+        entity = entity_observation.entity_id
 
         try:
-            if condition.state(self.hass, entity, entity_observation.get("to_state")):
+            if condition.state(self.hass, entity, entity_observation.to_state):
                 return True
         except ConditionError:
             return None
+        return None
 
     @property
     def extra_state_attributes(self):
         """Return the state attributes of the sensor."""
         attr_observations_list = [
-            obs.copy() for obs in self.current_observations.values() if obs is not None
+            observation.to_dict()
+            for observation in self.current_observations.values()
+            if observation is not None
         ]
 
-        for item in attr_observations_list:
-            item.pop("value_template", None)
-
         return {
-            ATTR_OBSERVATIONS: attr_observations_list,
+            ATTR_PROBABILITY: round(self.probability, 2),
+            ATTR_PROBABILITY_THRESHOLD: self._probability_threshold,
             ATTR_OCCURRED_OBSERVATION_ENTITIES: list(
```

Comment:
I'm not sure why a set is created, then converted to a list. May as well create a list directly:
```python
ATTR_OCCURRED_OBSERVATION_ENTITIES: [
                observation.entity_id
                for observation in self.current_observations.values()
                if observation is not None
                and observation.entity_id is not None
                and observation.observed is not None
            ],
```

Reply:
this causes a test to fail. Will try and work out why

Reply:
looks as if set functionality is needed to prevent duplication because a single entity can be can be in more than one observation
```
E       AssertionError: assert ['sensor.test...st_monitored'] == ['sensor.test_monitored']
E         Left contains one more item: 'sensor.test_monitored'
E         Full diff:
E         - ['sensor.test_monitored']
E         + ['sensor.test_monitored', 'sensor.test_monitored']
```

Reply:
I suggest that you add a comment to that effect then.

Reply:
Done